File: src/org/mtransit/parser/mt/data/MTripStop.java
package org.mtransit.parser.mt.data;

import java.util.List;

import org.mtransit.parser.Constants;

public class MTripStop implements Comparable<MTripStop> {

	private long tripId;
	private int stopId;
	private int stopSequence;
	private String uid;

	private boolean descentOnly = false;

	public MTripStop(long tripId, int stopId, int stopSequence) {
		this.tripId = tripId;
		this.stopId = stopId;
		this.stopSequence = stopSequence;
		this.uid = this.tripId + Constants.EMPTY + this.stopId;
	}

	public void setDescentOnly(boolean descentOnly) {
		this.descentOnly = descentOnly;
	}

	public boolean isDescentOnly() {
		return this.descentOnly;
	}

	public String getUID() {
		return this.uid;
	}

	public long getTripId() {
		return tripId;
	}

	public int getStopId() {
		return stopId;
	}

	public void setStopSequence(int stopSequence) {
		this.stopSequence = stopSequence;
	}

	@Override
	public boolean equals(Object obj) {
		MTripStop ts = (MTripStop) obj;
		if (ts.tripId != 0 && ts.tripId != tripId) {
			return false;
		}
		if (ts.stopId != 0 && ts.stopId != stopId) {
			return false;
		}
		if (ts.stopSequence != 0 && ts.stopSequence != stopSequence) {
			return false;
		}
		return true;
	}

	public boolean equalsExceptStopSequence(MTripStop ts) {
		if (ts.tripId != 0 && ts.tripId != tripId) {
			return false;
		}
		if (ts.stopId != 0 && ts.stopId != stopId) {
			return false;
		}
		return true;
	}

	@Override
	public String toString() {
		StringBuilder sb = new StringBuilder(); //
		sb.append(this.tripId); // TRIP ID
		sb.append(Constants.COLUMN_SEPARATOR); //
		sb.append(this.stopId); // STOP ID
		sb.append(Constants.COLUMN_SEPARATOR);
		sb.append(this.stopSequence); // STOP SEQUENCE
		sb.append(Constants.COLUMN_SEPARATOR);
		sb.append(this.descentOnly ? 1 : 0); // DESCENT ONLY
		return sb.toString();
	}

	@Override
	public int compareTo(MTripStop otherTripStop) {
		// sort by trip_id => stop_sequence
		if (this.tripId != otherTripStop.tripId) {
			return Long.compare(this.tripId, otherTripStop.tripId);
		}
		return stopSequence - otherTripStop.stopSequence;
	}

	public static String printStops(List<MTripStop> l) {
		StringBuilder sb = new StringBuilder();
		for (MTripStop mTripStop : l) {
			if (sb.length() > 0) {
				sb.append(Constants.COLUMN_SEPARATOR);
			}
			sb.append(mTripStop.stopId);
		}
		return sb.toString();
	}
}

File: src/main/java/org/mtransit/parser/mt/data/MTripStop.kt
package org.mtransit.parser.mt.data

import org.mtransit.parser.Constants

data class MTripStop(
    val tripId: Long,
    val stopId: Int,
    var stopSequence: Int,
    var isNoPickup: Boolean = false
) : Comparable<MTripStop> {

    // JAVA
    constructor(
        tripId: Long,
        stopId: Int,
        stopSequence: Int
    ) : this(
        tripId,
        stopId,
        stopSequence,
        false
    )

    val uID by lazy { getNewUID(tripId, stopId) }

    fun equalsExceptStopSequence(ts: MTripStop): Boolean {
        if (ts.tripId != 0L && ts.tripId != tripId) {
            return false
        }
        @Suppress("RedundantIf")
        if (ts.stopId != 0 && ts.stopId != stopId) {
            return false
        }
        return true
    }

    fun toFile() = buildString {
        append(tripId.toString()) // TRIP ID
        append(Constants.COLUMN_SEPARATOR) //
        append(stopId) // STOP ID
        append(Constants.COLUMN_SEPARATOR)
        append(stopSequence) // STOP SEQUENCE
        append(Constants.COLUMN_SEPARATOR)
        append(if (isNoPickup) 1 else 0) // DROP OFF ONLY
    }

    override fun compareTo(other: MTripStop): Int {
        // sort by trip_id => stop_sequence
        return if (tripId != other.tripId) {
            tripId.compareTo(other.tripId)
        } else stopSequence - other.stopSequence
    }

    @Suppress("unused")
    fun toStringSimple(): String {
        return "TS{$tripId>$stopId[$stopSequence](${if (this.isNoPickup) 0 else 1})"
    }

    @Suppress("unused")
    fun toStringSameTrip(): String {
        return "${this.stopSequence}:${this.stopId}";
    }

    companion object {

        @Suppress("unused")
        @JvmStatic
        fun containsStopIds(mainList: List<MTripStop>, otherList: List<MTripStop>): Boolean {
            return toStopIds(mainList).contains(toStopIds(otherList))
        }

        @Suppress("unused")
        @JvmStatic
        fun toStopIds(l: List<MTripStop>): String {
            return l.joinToString { "${it.stopId}" }
        }

        @Suppress("unused")
        @JvmStatic
        fun printTripStops(l: List<MTripStop>): String {
            return "[${l.size}] > ${l.joinToString { it.toStringSameTrip() }}"
        }

        @JvmStatic
        fun getNewUID(
            tripId: Long,
            stopId: Int
        ) = "${tripId}-${stopId}"
    }
}